marconi-tutorial/src/Marconi/Tutorial/Env.hs

```diff
@@ -0,0 +1,62 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Marconi.Tutorial.Env where
+
+import Cardano.BM.Trace (Trace)
+import Control.Concurrent (MVar)
+import Control.Concurrent.MVar (newEmptyMVar)
+import Control.Lens (makeLenses)
+import Data.IORef (IORef, newIORef)
+import Data.Text (Text)
+import Marconi.Tutorial.CLI qualified as CLI
+import Marconi.Tutorial.Indexers.AddressCount (AddressCountIndexer)
+
+data Env = Env
+  { _envIndexers :: IndexersEnv
+  -- ^ Access the Tutorial indexers for querying purposes.
+  , _envCliArgs :: !CLI.Options
+  , _envStdoutTrace :: !(Trace IO Text)
+  }
+
+-- | Should contain all the indexers required by the tutorial.
+newtype IndexersEnv = IndexersEnv
+  { _addressCountIndexerEnv :: AddressCountIndexerEnv
+  }
+
+newtype AddressCountIndexerEnv = AddressCountIndexerEnv
+  { _addressCountIndexerVar :: IORef (MVar AddressCountIndexer)
```

Comment:
It seems a bit ugly to me to have an `IORef` wrapping an `MVar`, since they're both mutable variables. I can see why you have to do it, because `Indexers.runIndexers` is creating the indexers as well as running them. However, perhaps there's a different approach that would eliminate this requirement as well as bringing other improvements.

One would be to separate the creation and running of the indexers, as we do in `marconi-sidechain` at present. Then the indexers can be placed into the environment when it's created. Combined with this, I think it's helpful to have different environments for the http server and the indexers, because they need to know about different things. For example, there's no need for the indexers to know which port the http server should be listening on. I've used this approach for the new RPC server.

Another approach would be to put the `MVar` directly into the environment, but leave it empty initially. `runIndexers` would then put the newly created indexer into the `MVar`, and then use the `MVar` in the same way that it does at present. In the long term, though, we might want to hide the `MVar` as an internal artifact of the indexer, like we do with the Utxo indexer in the new architecture.

Reply:
There's also a potential race condition in that the http server could start to query the `MVar` before the worker has been created and the real `MVar` has been rewritten into the `IORef`. Then, the http thread would hang forever because it's waiting on an `MVar` that will never be filled.

Reply:
Good idea! I'll go with this approach.